homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+from homeassistant.core import callback
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['mqtt']
+
+DEFAULT_SERVICES = SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_STOP |\
+                   SUPPORT_RETURN_HOME | SUPPORT_STATUS | SUPPORT_BATTERY |\
+                   SUPPORT_CLEAN_SPOT
+ALL_SERVICES = DEFAULT_SERVICES | SUPPORT_PAUSE | SUPPORT_LOCATE
+DEFAULT_SERVICE_STRINGS = services_to_strings(DEFAULT_SERVICES)
+
+DEFAULT_COMMAND_TOPIC = 'vacuum/command'
+DEFAULT_STATE_TOPIC = 'vacuum/state'
+
+DEFAULT_NAME = 'MQTT Vacuum'
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend({
+    vol.Optional(ATTR_SUPPORTED_FEATURES, default=DEFAULT_SERVICE_STRINGS):
+        vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(mqtt.CONF_COMMAND_TOPIC, default=DEFAULT_COMMAND_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(mqtt.CONF_STATE_TOPIC, default=DEFAULT_STATE_TOPIC):
+        mqtt.valid_publish_topic,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the vacuum."""
+    name = config.get(CONF_NAME)
+    command_topic = config.get(mqtt.CONF_COMMAND_TOPIC)
+    state_topic = config.get(mqtt.CONF_STATE_TOPIC)
+    supported_feature_strings = config.get(ATTR_SUPPORTED_FEATURES)
+    supported_features = strings_to_services(supported_feature_strings)
+    qos = config.get(mqtt.CONF_QOS)
+    state_template = cv.template('{{value_json.battery_level}}')
+    state_template.hass = hass
+
+    add_devices([
+        MqttVacuum(name, supported_features, command_topic, state_topic, qos,
+                   state_template),
+    ])
+
+
+class MqttVacuum(VacuumDevice):
+    """Representation of a MQTT-controlled vacuum."""
+
+    # pylint: disable=no-self-use
+    def __init__(self, name, supported_features, command_topic, state_topic,
+                 qos, state_template):
+        """Initialize the vacuum."""
+        self._name = name
+        self._command_topic = command_topic
+        self._state_topic = state_topic
+        self._supported_features = supported_features
+        self._qos = qos
+
+        self._payload_turn_on = "turn_on"
+        self._payload_turn_off = "turn_off"
+        self._payload_return_to_base = "return_to_base"
+        self._payload_stop = "stop"
+        self._payload_clean_spot = "clean_spot"
+        self._payload_locate = "locate"
+        self._payload_start_pause = "start_pause"
+        self._retain = False
+
+        self._state_template = state_template
+
+        self._state = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+            try:
+                payload = json.loads(payload)
```

Comment:
I think that need to be more generic like other mqtt component. Maybe wuth templates (that support also json) and variable topics for that points. In yiur case you can set all the sane topic and use the tempkate to parse the correct value on it.

Reply:
Ok sure, will do